src/uwtools/config/tools.py

```diff
@@ -38,6 +37,57 @@ def compare_configs(
     return cfg_a.compare_config(cfg_b.data)
 
 
+def config_check_depths_dump(config_obj: dict, target_format: str) -> None:
+    """
+    Check that the depth of the input config does not exceed the target format's max.
+
+    :param config_obj: The reference config object.
+    :param target_format: The target format.
+    :raises: UWConfigError on excessive input-config depth.
+    """
+    cfgclass = format_to_config(target_format)
+    if cfgclass.DEPTH and depth(config_obj) > cfgclass.DEPTH:
+        msg = "Invalid depth-%s input for type-'%s' config" % (
+            depth(config_obj),
+            cfgclass.depth,
+        )
+        raise log_and_error(msg)
+
+
+def config_check_depths_realize(config_obj: Config, target_format: str) -> None:
+    """
+    Check that the depth of the input config does not exceed the target format's max.
+
+    :param config_obj: The reference config object.
+    :param target_format: The target format.
+    :raises: UWConfigError on excessive input-config depth.
+    """
+    cfgclass = format_to_config(target_format)
+    if cfgclass.DEPTH and config_obj.depth != cfgclass.DEPTH:
+        msg = "Cannot write depth-%s input to type-'%s' config" % (
+            config_obj.depth,
+            cfgclass.depth,
+        )
+        raise log_and_error(msg)
+
+
+def config_check_depths_update(config_obj: Config, target_format: str) -> None:
+    """
+    Check that the depth of the input config does not exceed the target format's max.
+
+    :param config_obj: The reference config object.
+    :param target_format: The target format.
+    :raises: UWConfigError on excessive input-config depth.
+    """
+    cfgclass = format_to_config(target_format)
+    if cfgclass.DEPTH and config_obj.depth > cfgclass.DEPTH:
+        msg = "Cannot update depth-%s input with type-'%s' values" % (
+            config_obj.depth,
+            cfgclass.depth,
+        )
+        raise log_and_error(msg)
```

Comment:
This is all looking better to me than the big block of conditionals!

I'm curious as to why one of these functions accepts a `dict` and the others accept `Config`s. Wouldn't either input be an appropriate thing to test? If so, we could take a `Union[dict, Config]` argument and then e.g.
```
config_obj = config_obj.data if isinstance(config_obj, Config) else config_obj
```
to ensure that what follows operates strictly on a `dict`. That is: Establish an invariant and then proceed with confidence.

I feel like that'd make these functions more universally useful throughout the codebase, and would eliminate an apparently (but please correct me if I'm wrong) unnecessary difference.

Reply:
This is kind of an interesting structural question - since `dump` is a concrete static method, it takes the in-memory object and not anything else. In the first version I left the `formats` with individual `assert` statements, then the second version had everything together with that `Union` logic

Reply:
Ah wait, I have an idea, I think this might be the way but for another reason, one sec

Reply:
There, got it - it turns out that while we were checking `dump_dict()`, `dump()` was not being checked. While unlikely to be broken, it doesn't hurt and makes a good direct use for returning `Union`

Reply:
Looking good!

At this point, `config_check_depths_dump()` and `config_check_depths_update()` are identical except for a few characters in their respective `msg` values. Those two functions could be collapsed into a single line each that calls a common helper that takes the differing part of the `msg` as a 3rd argument. To cut down on boilerplate, I think the official docstring could live in the helper, with the two gutted functions just saying `See <whatever>()'s docstring.` where `<whatever>` is whatever the helper is called. We're honing in on the essentials now, trimming away the accidentals...

As a potential final refactor, consider that `config_check_depths_realize()` differs from the other two only in the comparison it performs (and its `msg`, but the paragraph above would generalize that). That, too, could be factored out and passed in to the common helper. For example, if `config_check_depths_realize()` defined
```
bad_depth = lambda need, have: need and have != need
```
and the other two defined
```
bad_depth = lambda need, have: need and have > need
```
and passed `bad_depth` in to the proposed helper as a 4th argument, and where, say, `action` is the 3rd argument from above making `msg` conext-appropriate, the helper could have e.g.
```
if bad_depth(cfgclass.DEPTH, config_obj.depth):
    raise log_and_error(
        "Cannot %s depth-%s config to type-'%s' config" % (
            action, config_obj.depth, target_format
        )
    )
```
(The type of a lambda/function/method argument is `Callable`, which can be imported from `typing`.)

That might be a very clean and DRY refactor.

Reply:
Took me a minute to follow, but this is a brilliantly clean solution. I need to consider lambdas more!